native/spark-expr/src/map_funcs/map_filter.rs

```diff
@@ -0,0 +1,539 @@
+                        // If filtering produces multiple maps, return as array
+                        Ok(ColumnarValue::Array(Arc::new(filtered)))
+                    }
+                }
+                ScalarValue::Null => {
+                    // Handle null scalar map
+                    Ok(ColumnarValue::Scalar(ScalarValue::Null))
+                }
+                _ => Err(DataFusionError::Internal(
+                    "Invalid map data received".to_string()
+                )),
+            }
+        }
+    }
+}
+
+fn filter_map_with_lambda(map_array: &MapArray, lambda_arg: &ColumnarValue) -> DataFusionResult<MapArray> {
+    let lambda_expr = extract_lambda_expression(lambda_arg)?;
+    let entries = map_array.entries();
+    let entries_str = entries.as_any().downcast_ref::<StructArray>().ok_or_else(||
+        DataFusionError::Internal("Expected StructArray for map entries".to_string()))?;
+
+    let keys = entries_str.column(0);
+    let values = entries_str.column(1);
+    let filter_mask = evaluate_lambda_on_pairs(&lambda_expr, keys, values)?;
+
+    let filtered_entries = filter_struct_array(entries_str, &filter_mask)?;
+    let offsets = compute_filtered_offsets(map_array, &filter_mask)?;
+    let filtered_map = match map_array.data_type() {
+        DataType::Map(field, _) => field.clone(),
+        _ => return Err(DataFusionError::Internal(
+            "Invalid map datatype".to_string())),
+    };
+
+    MapArray::try_new(filtered_map, offsets, filtered_entries, map_array.nulls()
+        .cloned(), false,).map_err(|e| DataFusionError::Internal(format!("Arrow error: {}", e)))
+}
+
+fn extract_lambda_expression(lambda_arg: &ColumnarValue) -> DataFusionResult<String> {
+    // Handle string based lambda expressions
+    match lambda_arg {
+        ColumnarValue::Scalar(scalar_value) => {
+            match scalar_value {
+                ScalarValue::Utf8(Some(expr)) => Ok(expr.clone()),
+                _ => Err(DataFusionError::Internal("Lambda expression must be a
+  string".to_string())),
+            }
+        }
+        _ => Err(DataFusionError::Internal("Lambda expression must be a scalar
+  string".to_string())),
+    }
+}
+
+fn evaluate_lambda_on_pairs(
+    lambda_expr: &str,
+    keys: &ArrayRef,
+    values: &ArrayRef
+) -> DataFusionResult<BooleanArray> {
+    let mut results = Vec::new();
+    let len = keys.len();
+
+    for i in 0..len {
+        let result = evaluate_lambda_comparison(lambda_expr, keys, values, i)?;
+        results.push(result);
+    }
+
+    Ok(BooleanArray::from(results))
+}
+
+fn evaluate_lambda_comparison(
+    lambda_expr: &str,
+    keys: &ArrayRef,
+    values: &ArrayRef,
+    index: usize
+) -> DataFusionResult<Option<bool>> {
+    // Handle null values
+    if keys.is_null(index) || values.is_null(index) {
+        return Ok(Some(false)); // Spark behavior: nulls are filtered out
+    }
+
+    // Parse the lambda expression
+    if lambda_expr.contains(" >= ") {
```

Comment:
Is this parsing sufficient? How complex can they be? I'm surprised they're coming out of Spark as strings and not parsed already into some sort of expression.